snippets upload does not work on Windows

**Describe the bug**
A clear and concise description of what the bug is.

I was following the online documentation (https://registry.terraform.io/providers/bpg/proxmox/latest/docs/resources/virtual_environment_file) for creating the cloud-config file.  

```
resource "proxmox_virtual_environment_file" "cloud_config_1" {
  content_type = "snippets"
  datastore_id = "local"
  node_name    = "proxmox-01"


  source_raw {
    data = <<EOF
#cloud-config
chpasswd:
  list: |
    ubuntu:${random_password.ubuntu_vm_password.result}
  expire: false
packages:
  - qemu-guest-agent
  - python3
  - python3-pip
users:
  - default
  - name: ubuntu
    groups: sudo
    shell: /bin/bash
    ssh-authorized-keys:
      - "<redacted>"
      - "<redacted>"
    sudo: ALL=(ALL) NOPASSWD:ALL
EOF

    file_name = "my.cloud-config.yaml"
  }
}

```
However, when executing this I get the error: Error: failed to create file /tmp/tfpve/e802c02a-ce68-4ce3-af10-fdc9ff6c753f/snippets/my.cloud-config.yaml: file does not exist.  I go to the server, and I see this:

```
$ ls
'\tmp\tfpve\afac11a0-962f-4746-b525-59801be6ef96\snippets'  '\tmp\tfpve\e802c02a-ce68-4ce3-af10-fdc9ff6c753f\snippets'   test.txt

```

So of course it can't find them...they are there in the wrong pathing format.  So am I doing this wrong or?

**To Reproduce**
Steps to reproduce the behavior:

Dies as soon as I attempt to run the terraform with the above error.


**Expected behavior**
A clear and concise description of what you expected to happen.

Create the cloud config file.  


**Additional context**
Add any other context about the problem here.

- Single or clustered Proxmox: clustered
- Provider version (ideally it should be the latest version): 8.1.4
- Terraform version: 1.7.3
- OS (where you run Terraform from): windows
- Debug logs (`TF_LOG=DEBUG terraform apply`):

```
proxmox_virtual_environment_file.cloud_config: Creating...
2024-02-27T12:36:45.493-0600 [INFO]  Starting apply for proxmox_virtual_environment_file.cloud_config
2024-02-27T12:36:45.493-0600 [DEBUG] proxmox_virtual_environment_file.cloud_config: applying the planned Create change
2024-02-27T12:36:45.493-0600 [DEBUG] provider.terraform-provider-proxmox_v0.46.4.exe: Sending HTTP Request: tf_http_req_body="" tf_http_req_version=HTTP/1.1 tf_http_trans_id=57f696d7-7dcc-768f-781e-1b523b163e85 @caller=github.com/hashicorp/terraform-plugin-sdk/v2@v2.32.0/helper/logging/logging_http_transport.go:162 Accept-Encoding=gzip User-Agent=Go-http-client/1.1 tf_mux_provider=tf5to6server.v5tov6Server tf_resource_type=proxmox_virtual_environment_file Authorization="PVEAPIToken=terraform@pam!provider=3ed9ae1e-060b-4091-8775-ae17b0a314ec" tf_http_op_type=request tf_http_req_method=GET tf_http_req_uri=/api2/json/nodes/leen-proxmox-01/storage/local/content @module=proxmox Accept=application/json Host=<redacted>:8006 tf_provider_addr=registry.terraform.io/bpg/proxmox tf_req_id=09f80f3a-931d-2bd6-a434-f2a5ca2bd016 tf_rpc=ApplyResourceChange timestamp=2024-02-27T12:36:45.493-0600
2024-02-27T12:36:45.684-0600 [DEBUG] provider.terraform-provider-proxmox_v0.46.4.exe: Received HTTP Response: tf_http_res_body="{\"data\":[{\"ctime\":1708374984,\"volid\":\"local:iso/en-us_windows_11_business_editions_version_21h2_updated_feb_2023_x64_dvd_ddb0ab83.iso\",\"content\":\"iso\",\"format\":\"iso\",\"size\":5842262016},{\"content\":\"iso\",\"format\":\"iso\",\"size\":5550684160,\"volid\":\"local:iso/en-us_windows_server_2022_x64_dvd_620d7eac.iso\",\"ctime\":1708374984},{\"content\":\"iso\",\"format\":\"iso\",\"size\":6006587392,\"volid\":\"local:iso/en_windows_server_2016_updated_feb_2018_x64_dvd_11636692.iso\",\"ctime\":1708374984},{\"content\":\"iso\",\"size\":5569699840,\"format\":\"iso\",\"volid\":\"local:iso/en_windows_server_2019_updated_march_2021_x64_dvd_ec2626a1.iso\",\"ctime\":1708374984},{\"ctime\":1708374984,\"volid\":\"local:iso/pfSense-CE-2.7.2-RELEASE-amd64.iso\",\"format\":\"iso\",\"content\":\"iso\",\"size\":874672128},{\"ctime\":1708374984,\"volid\":\"local:iso/ubuntu-22.04.3-desktop-amd64.iso\",\"format\":\"iso\",\"content\":\"iso\",\"size\":5037662208},{\"volid\":\"local:iso/ubuntu.img\",\"size\":648741376,\"content\":\"iso\",\"format\":\"iso\",\"ctime\":1709050453},{\"ctime\":1708478875,\"size\":627519488,\"content\":\"iso\",\"format\":\"iso\",\"volid\":\"local:iso/virtio-win-0.1.240.iso\"},{\"content\":\"snippets\",\"format\":\"snippet\",\"size\":477,\"volid\":\"local:snippets/leen.cloud-config.yaml\",\"ctime\":1709056989}]}" tf_http_res_status_code=200 tf_provider_addr=registry.terraform.io/bpg/proxmox tf_rpc=ApplyResourceChange @caller=github.com/hashicorp/terraform-plugin-sdk/v2@v2.32.0/helper/logging/logging_http_transport.go:162 Content-Type=application/json;charset=UTF-8 Expires="Tue, 27 Feb 2024 18:36:46 GMT" tf_http_op_type=response tf_http_trans_id=57f696d7-7dcc-768f-781e-1b523b163e85 tf_req_id=09f80f3a-931d-2bd6-a434-f2a5ca2bd016 tf_resource_type=proxmox_virtual_environment_file @module=proxmox Connection=[Keep-Alive, Keep-Alive] Date="Tue, 27 Feb 2024 18:36:46 GMT" Pragma=no-cache tf_mux_provider=tf5to6server.v5tov6Server Cache-Control=max-age=0 Server=pve-api-daemon/3.0 tf_http_res_status_reason="200 OK" tf_http_res_version=HTTP/1.1 timestamp=2024-02-27T12:36:45.684-0600
2024-02-27T12:36:45.685-0600 [DEBUG] provider.terraform-provider-proxmox_v0.46.4.exe: Sending HTTP Request: Accept=application/json tf_mux_provider=tf5to6server.v5tov6Server tf_provider_addr=registry.terraform.io/bpg/proxmox tf_resource_type=proxmox_virtual_environment_file tf_rpc=ApplyResourceChange @module=proxmox Host=<redacted>:8006 User-Agent=Go-http-client/1.1 tf_req_id=09f80f3a-931d-2bd6-a434-f2a5ca2bd016 @caller=github.com/hashicorp/terraform-plugin-sdk/v2@v2.32.0/helper/logging/logging_http_transport.go:162 Accept-Encoding=gzip Authorization="PVEAPIToken=terraform@pam!provider=3ed9ae1e-060b-4091-8775-ae17b0a314ec" tf_http_req_body="" tf_http_req_uri=/api2/json/storage/local tf_http_req_version=HTTP/1.1 tf_http_trans_id=67539e0d-ea99-7ec3-88c1-a87978167400 tf_http_op_type=request tf_http_req_method=GET timestamp=2024-02-27T12:36:45.685-0600
2024-02-27T12:36:45.748-0600 [DEBUG] provider.terraform-provider-proxmox_v0.46.4.exe: Received HTTP Response: tf_http_op_type=response tf_http_res_body="{\"data\":{\"content\":\"vztmpl,rootdir,snippets,images,backup,iso\",\"path\":\"/var/lib/vz\",\"type\":\"dir\",\"shared\":0,\"is_mountpoint\":\"no\",\"digest\":\"a8e1eef050b0574f808b8e6956a3dbfdbd67fc1e\",\"storage\":\"local\"}}" tf_http_res_status_code=200 @module=proxmox Content-Length=200 tf_mux_provider=tf5to6server.v5tov6Server tf_provider_addr=registry.terraform.io/bpg/proxmox @caller=github.com/hashicorp/terraform-plugin-sdk/v2@v2.32.0/helper/logging/logging_http_transport.go:162 Cache-Control=max-age=0 Connection=Keep-Alive Pragma=no-cache Server=pve-api-daemon/3.0 tf_http_trans_id=67539e0d-ea99-7ec3-88c1-a87978167400 tf_resource_type=proxmox_virtual_environment_file tf_rpc=ApplyResourceChange Date="Tue, 27 Feb 2024 18:36:47 GMT" Expires="Tue, 27 Feb 2024 18:36:47 GMT" tf_http_res_status_reason="200 OK" tf_http_res_version=HTTP/1.1 tf_req_id=09f80f3a-931d-2bd6-a434-f2a5ca2bd016 Content-Type=application/json;charset=UTF-8 timestamp=2024-02-27T12:36:45.747-0600
2024-02-27T12:36:45.748-0600 [DEBUG] provider.terraform-provider-proxmox_v0.46.4.exe: uploading file to the node datastore using SFTP: remote_dir=/tmp/tfpve/3a2536bf-86ed-437d-ba70-8e58cc7b15e5 tf_resource_type=proxmox_virtual_environment_file content_type=snippets node_address="map[Address:<redacted> Port:22]" tf_mux_provider=tf5to6server.v5tov6Server tf_req_id=09f80f3a-931d-2bd6-a434-f2a5ca2bd016 @module=proxmox file_name=example.cloud-config.yaml tf_provider_addr=registry.terraform.io/bpg/proxmox tf_rpc=ApplyResourceChange @caller=github.com/bpg/terraform-provider-proxmox/proxmox/ssh/client.go:162 timestamp=2024-02-27T12:36:45.748-0600
2024-02-27T12:36:45.748-0600 [INFO]  provider.terraform-provider-proxmox_v0.46.4.exe: agent is set to false: @module=proxmox tf_mux_provider=tf5to6server.v5tov6Server tf_provider_addr=registry.terraform.io/bpg/proxmox tf_req_id=09f80f3a-931d-2bd6-a434-f2a5ca2bd016 @caller=github.com/bpg/terraform-provider-proxmox/proxmox/ssh/client.go:303 tf_resource_type=proxmox_virtual_environment_file tf_rpc=ApplyResourceChange timestamp=2024-02-27T12:36:45.748-0600
2024-02-27T12:36:46.213-0600 [DEBUG] provider.terraform-provider-proxmox_v0.46.4.exe: SSH connection established: host=<redacted>:22 tf_req_id=09f80f3a-931d-2bd6-a434-f2a5ca2bd016 tf_resource_type=proxmox_virtual_environment_file @caller=github.com/bpg/terraform-provider-proxmox/proxmox/ssh/client.go:398 @module=proxmox user=terraform tf_mux_provider=tf5to6server.v5tov6Server tf_provider_addr=registry.terraform.io/bpg/proxmox tf_rpc=ApplyResourceChange timestamp=2024-02-27T12:36:46.213-0600
2024-02-27T12:36:47.089-0600 [ERROR] provider.terraform-provider-proxmox_v0.46.4.exe: Response contains error diagnostic: tf_proto_version=6.4 tf_provider_addr=registry.terraform.io/bpg/proxmox @caller=github.com/hashicorp/terraform-plugin-go@v0.21.0/tfprotov6/internal/diag/diagnostics.go:62 tf_req_id=09f80f3a-931d-2bd6-a434-f2a5ca2bd016 tf_resource_type=proxmox_virtual_environment_file diagnostic_summary="failed to create file /tmp/tfpve/3a2536bf-86ed-437d-ba70-8e58cc7b15e5/snippets/example.cloud-config.yaml: file does not exist" tf_rpc=ApplyResourceChange @module=sdk.proto diagnostic_detail="" diagnostic_severity=ERROR timestamp=2024-02-27T12:36:47.089-0600
2024-02-27T12:36:47.110-0600 [DEBUG] State storage *statemgr.Filesystem declined to persist a state snapshot
2024-02-27T12:36:47.110-0600 [ERROR] vertex "proxmox_virtual_environment_file.cloud_config" error: failed to create file /tmp/tfpve/3a2536bf-86ed-437d-ba70-8e58cc7b15e5/snippets/example.cloud-config.yaml: file does not exist
╷
│ Warning: Resource targeting is in effect
│
│ You are creating a plan with the -target option, which means that the result of this plan may not represent all of the changes requested by the current configuration.
│
│ The -target option is not for routine use, and is provided only for exceptional situations such as recovering from errors or mistakes, or when Terraform specifically suggests to use it as part of an error message.
╵
╷
│ Warning: Applied changes may be incomplete
│
│ The plan was created with the -target option in effect, so some changes requested in the configuration may have been ignored and the output values may not be fully updated. Run the following command to verify that no other changes are pending:
│     terraform plan
│
│ Note that the -target option is not suitable for routine use, and is provided only for exceptional situations such as recovering from errors or mistakes, or when Terraform specifically suggests to use it as part of an error message.
╵
╷
│ Error: failed to create file /tmp/tfpve/3a2536bf-86ed-437d-ba70-8e58cc7b15e5/snippets/example.cloud-config.yaml: file does not exist
│
│   with proxmox_virtual_environment_file.cloud_config,
│   on main.tf line 45, in resource "proxmox_virtual_environment_file" "cloud_config":
│   45: resource "proxmox_virtual_environment_file" "cloud_config" {
│
╵
2024-02-27T12:36:47.122-0600 [DEBUG] provider.stdio: received EOF, stopping recv loop: err="rpc error: code = Unavailable desc = error reading from server: EOF"
2024-02-27T12:36:47.144-0600 [DEBUG] provider: plugin process exited

--- a/.github/workflows/test.yml
+++ b/.github/workflows/test.yml
@@ -77,7 +77,7 @@ jobs:
         run: make docs && git diff --exit-code
 
   testacc:
-    if: "!contains(github.head_ref, 'renovate/')"
+    if: "!contains(github.head_ref, 'renovate/') && !contains(github.head_ref, 'release-please')"
     name: Dispatch Acceptance Tests
     needs: build
     runs-on: ubuntu-latest

--- a/.github/workflows/test.yml
+++ b/.github/workflows/test.yml
@@ -78,12 +78,9 @@ jobs:
     needs: build
     runs-on: ubuntu-latest
     steps:
-      - name: Dump GitHub context
-        env:
-          GITHUB_CONTEXT: ${{ toJson(github) }}
-        run: echo "$GITHUB_CONTEXT"
-
       - name: Invoke acceptance tests workflow
         uses: benc-uk/workflow-dispatch@v1
         with:
           workflow: testacc.yml
+          ref: "main"
+          inputs: '{"ref": "${{ github.head_ref }}" }'

--- a/.github/workflows/test.yml
+++ b/.github/workflows/test.yml
@@ -8,6 +8,7 @@ on:
 
 jobs:
   build:
+    name: Build
     runs-on: ubuntu-latest
     timeout-minutes: 5
     steps:
@@ -40,6 +41,7 @@ jobs:
         run: go vet . && go build -v .
 
   test:
+    name: Unit Tests
     needs: build
     runs-on: ubuntu-latest
     steps:
@@ -75,6 +77,7 @@ jobs:
         run: make docs && git diff --exit-code
 
   testacc:
+    name: Dispatch Acceptance Tests
     needs: build
     runs-on: ubuntu-latest
     steps:
@@ -82,5 +85,5 @@ jobs:
         uses: benc-uk/workflow-dispatch@v1
         with:
           workflow: testacc.yml
-          ref: "main"
+          ref: ${{ github.event.pull_request.head.ref }}
           inputs: '{"ref": "${{ github.head_ref }}" }'

--- a/.github/workflows/test.yml
+++ b/.github/workflows/test.yml
@@ -77,6 +77,7 @@ jobs:
         run: make docs && git diff --exit-code
 
   testacc:
+    if: "!contains(github.head_ref, 'renovate/')"
     name: Dispatch Acceptance Tests
     needs: build
     runs-on: ubuntu-latest

--- a/.github/workflows/testacc.yml
+++ b/.github/workflows/testacc.yml
@@ -11,6 +11,7 @@ on:
 jobs:
   acceptance:
     strategy:
+      max-parallel: 1
       matrix:
         os: [ ubuntu-latest, windows-latest, macos-latest ]
         terraform: [ 1.6 ]

--- a/.github/workflows/testacc.yml
+++ b/.github/workflows/testacc.yml
@@ -14,7 +14,7 @@ jobs:
     strategy:
       max-parallel: 1
       matrix:
-        os: [ ubuntu-latest ]
+        os: [ ubuntu-latest, windows-latest, macos-latest ]
         terraform: [ 1.6 ]
     runs-on: ${{ matrix.os }}
     environment: pve-acc

--- a/.github/workflows/testacc.yml
+++ b/.github/workflows/testacc.yml
@@ -12,7 +12,6 @@ jobs:
   acceptance:
     #refs/heads/renovate/tools
     strategy:
-      max-parallel: 1
       matrix:
 #        os: [ ubuntu-latest, windows-latest, macos-latest ]
         os: [ windows-latest ]

--- a/.github/workflows/testacc.yml
+++ b/.github/workflows/testacc.yml
@@ -10,11 +10,9 @@ on:
 
 jobs:
   acceptance:
-    #refs/heads/renovate/tools
     strategy:
       matrix:
-#        os: [ ubuntu-latest, windows-latest, macos-latest ]
-        os: [ windows-latest ]
+        os: [ ubuntu-latest, windows-latest, macos-latest ]
         terraform: [ 1.6 ]
     runs-on: ${{ matrix.os }}
     environment: pve-acc

--- a/.github/workflows/testacc.yml
+++ b/.github/workflows/testacc.yml
@@ -14,7 +14,8 @@ jobs:
     strategy:
       max-parallel: 1
       matrix:
-        os: [ ubuntu-latest, windows-latest, macos-latest ]
+#        os: [ ubuntu-latest, windows-latest, macos-latest ]
+        os: [ windows-latest ]
         terraform: [ 1.6 ]
     runs-on: ${{ matrix.os }}
     environment: pve-acc

--- a/docs/index.md
+++ b/docs/index.md
@@ -200,7 +200,7 @@ You can configure the `sudo` privilege for the user via the command line on the 
     ```sh
     terraform ALL=(root) NOPASSWD: /sbin/pvesm
     terraform ALL=(root) NOPASSWD: /sbin/qm
-    terraform ALL=(root) NOPASSWD: /usr/bin/mv /tmp/tfpve/* /var/lib/vz/*
+    terraform ALL=(root) NOPASSWD: /usr/bin/tee /var/lib/vz/*
     ```
 
   Save the file and exit.

--- a/example/main.tf
+++ b/example/main.tf
@@ -1,9 +1,9 @@
 provider "proxmox" {
   endpoint = var.virtual_environment_endpoint
-  username = var.virtual_environment_username
-  password = var.virtual_environment_password
+  api_token = var.virtual_environment_api_token
   insecure = true
   ssh {
     agent = true
+    username = var.virtual_environment_ssh_username
   }
 }

--- a/example/variables.tf
+++ b/example/variables.tf
@@ -3,12 +3,12 @@ variable "virtual_environment_endpoint" {
   description = "The endpoint for the Proxmox Virtual Environment API (example: https://host:port)"
 }
 
-variable "virtual_environment_password" {
+variable "virtual_environment_api_token" {
   type        = string
-  description = "The password for the Proxmox Virtual Environment API"
+  description = "The API token for the Proxmox Virtual Environment API"
 }
 
-variable "virtual_environment_username" {
+variable "virtual_environment_ssh_username" {
   type        = string
-  description = "The username and realm for the Proxmox Virtual Environment API (example: root@pam)"
+  description = "The username for the Proxmox Virtual Environment API"
 }

--- a/fwprovider/tests/resource_file_test.go
+++ b/fwprovider/tests/resource_file_test.go
@@ -211,7 +211,7 @@ resource "proxmox_virtual_environment_file" "test" {
   }
   %s
 }
-	`, getProviderConfig(t), accTestNodeName, fname, strings.Join(extra, "\n"))
+	`, getProviderConfig(t), accTestNodeName, strings.ReplaceAll(fname, `\`, `/`), strings.Join(extra, "\n"))
 }
 
 func testAccResourceFileTwoSourcesCreatedConfig(t *testing.T) string {
@@ -274,7 +274,7 @@ resource "proxmox_virtual_environment_file" "test" {
     path = "%s"
   }
 }
-	`, getProviderConfig(t), accTestNodeName, fname)
+	`, getProviderConfig(t), accTestNodeName, strings.ReplaceAll(fname, `\`, `/`))
 }
 
 func testAccResourceFileSnippetUpdatedCheck(fname string) resource.TestCheckFunc {

--- a/fwprovider/tests/resource_file_test.go
+++ b/fwprovider/tests/resource_file_test.go
@@ -146,20 +146,12 @@ func uploadSnippetFile(t *testing.T, file *os.File) {
 	defer f.Close()
 
 	fname := filepath.Base(file.Name())
-	err = sshClient.NodeUpload(context.Background(), "pve", "/tmp/tfpve/testacc",
+	err = sshClient.NodeStreamUpload(context.Background(), "pve", "/var/lib/vz/",
 		&api.FileUploadRequest{
 			ContentType: "snippets",
 			FileName:    fname,
 			File:        f,
 		})
-	require.NoError(t, err)
-
-	_, err = sshClient.ExecuteNodeCommands(context.Background(), "pve", []string{
-		fmt.Sprintf(`%s; try_sudo "mv /tmp/tfpve/testacc/snippets/%s /var/lib/vz/snippets/%s" && rm -rf /tmp/tfpve/testacc/`,
-			ssh.TrySudo,
-			fname, fname,
-		),
-	})
 	require.NoError(t, err)
 }
 

--- a/fwprovider/tests/resource_file_test.go
+++ b/fwprovider/tests/resource_file_test.go
@@ -125,11 +125,12 @@ func uploadSnippetFile(t *testing.T, file *os.File) {
 	u, err := url.ParseRequestURI(endpoint)
 	require.NoError(t, err)
 
+	sshAgent := utils.GetAnyBoolEnv("PROXMOX_VE_SSH_AGENT")
 	sshUsername := utils.GetAnyStringEnv("PROXMOX_VE_SSH_USERNAME")
-	sshAgentSocket := utils.GetAnyStringEnv("SSH_AUTH_SOCK", "PROXMOX_VE_SSH_AUTH_SOCK", "PM_VE_SSH_AUTH_SOCK")
+	sshAgentSocket := utils.GetAnyStringEnv("SSH_AUTH_SOCK", "PROXMOX_VE_SSH_AUTH_SOCK")
 	sshPrivateKey := utils.GetAnyStringEnv("PROXMOX_VE_SSH_PRIVATE_KEY")
 	sshClient, err := ssh.NewClient(
-		sshUsername, "", true, sshAgentSocket, sshPrivateKey,
+		sshUsername, "", sshAgent, sshAgentSocket, sshPrivateKey,
 		"", "", "",
 		&nodeResolver{
 			node: ssh.ProxmoxNode{

--- a/proxmox/ssh/client.go
+++ b/proxmox/ssh/client.go
@@ -300,6 +300,7 @@ func (c *client) NodeStreamUpload(
 	if err != nil {
 		return fmt.Errorf("failed to create SSH session: %w", err)
 	}
+
 	defer func(session *ssh.Session) {
 		e := session.Close()
 		if e != nil {

--- a/proxmox/ssh/client.go
+++ b/proxmox/ssh/client.go
@@ -51,6 +51,10 @@ type Client interface {
 	// NodeUpload uploads a file to a node.
 	NodeUpload(ctx context.Context, nodeName string,
 		remoteFileDir string, fileUploadRequest *api.FileUploadRequest) error
+
+	// NodeStreamUpload uploads a file to a node by streaming its content over SSH.
+	NodeStreamUpload(ctx context.Context, nodeName string,
+		remoteFileDir string, fileUploadRequest *api.FileUploadRequest) error
 }
 
 type client struct {
@@ -234,6 +238,111 @@ func (c *client) NodeUpload(
 		return fmt.Errorf("failed to upload file %s: %w", remoteFilePath, err)
 	}
 
+	if bytesUploaded != fileSize {
+		return fmt.Errorf("failed to upload file %s: uploaded %d bytes, expected %d bytes",
+			remoteFilePath, bytesUploaded, fileSize)
+	}
+
+	tflog.Debug(ctx, "uploaded file to datastore", map[string]interface{}{
+		"remote_file_path": remoteFilePath,
+		"size":             bytesUploaded,
+	})
+
+	return nil
+}
+
+func (c *client) NodeStreamUpload(
+	ctx context.Context,
+	nodeName string,
+	remoteFileDir string,
+	d *api.FileUploadRequest,
+) error {
+	ip, err := c.nodeResolver.Resolve(ctx, nodeName)
+	if err != nil {
+		return fmt.Errorf("failed to find node endpoint: %w", err)
+	}
+
+	tflog.Debug(ctx, "uploading file to the node datastore via SSH input stream ", map[string]interface{}{
+		"node_address": ip,
+		"remote_dir":   remoteFileDir,
+		"file_name":    d.FileName,
+		"content_type": d.ContentType,
+	})
+
+	fileInfo, err := d.File.Stat()
+	if err != nil {
+		return fmt.Errorf("failed to get file info: %w", err)
+	}
+
+	fileSize := fileInfo.Size()
+
+	sshClient, err := c.openNodeShell(ctx, ip)
+	if err != nil {
+		return fmt.Errorf("failed to open SSH client: %w", err)
+	}
+
+	defer func(sshClient *ssh.Client) {
+		e := sshClient.Close()
+		if e != nil {
+			tflog.Error(ctx, "failed to close SSH client", map[string]interface{}{
+				"error": e,
+			})
+		}
+	}(sshClient)
+
+	if d.ContentType != "" {
+		remoteFileDir = filepath.Join(remoteFileDir, d.ContentType)
+	}
+
+	remoteFilePath := strings.ReplaceAll(filepath.Join(remoteFileDir, d.FileName), `\`, "/")
+
+	sshSession, err := sshClient.NewSession()
+	if err != nil {
+		return fmt.Errorf("failed to create SSH session: %w", err)
+	}
+	defer func(session *ssh.Session) {
+		e := session.Close()
+		if e != nil {
+			tflog.Error(ctx, "failed to close SSH session", map[string]interface{}{
+				"error": e,
+			})
+		}
+	}(sshSession)
+
+	sshSession.Stdin = d.File
+
+	output, err := sshSession.CombinedOutput(
+		fmt.Sprintf(`%s; try_sudo "/usr/bin/tee %s"`, TrySudo, remoteFilePath),
+	)
+	if err != nil {
+		return fmt.Errorf("error transferring file: %s", string(output))
+	}
+
+	sftpClient, err := sftp.NewClient(sshClient)
+	if err != nil {
+		return fmt.Errorf("failed to create SFTP client: %w", err)
+	}
+
+	defer func(sftpClient *sftp.Client) {
+		e := sftpClient.Close()
+		if e != nil {
+			tflog.Error(ctx, "failed to close SFTP client", map[string]interface{}{
+				"error": e,
+			})
+		}
+	}(sftpClient)
+
+	remoteFile, err := sftpClient.Open(remoteFilePath)
+	if err != nil {
+		return fmt.Errorf("failed to open remote file %s: %w", remoteFilePath, err)
+	}
+
+	remoteStat, err := remoteFile.Stat()
+	if err != nil {
+		return fmt.Errorf("failed to read remote file %s: %w", remoteFilePath, err)
+	}
+
+	bytesUploaded := remoteStat.Size()
 	if bytesUploaded != fileSize {
 		return fmt.Errorf("failed to upload file %s: uploaded %d bytes, expected %d bytes",
 			remoteFilePath, bytesUploaded, fileSize)

--- a/proxmoxtf/resource/file.go
+++ b/proxmoxtf/resource/file.go
@@ -22,7 +22,6 @@ import (
 	"strings"
 	"time"
 
-	"github.com/google/uuid"
 	"github.com/hashicorp/go-cty/cty"
 	"github.com/hashicorp/terraform-plugin-log/tflog"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
@@ -573,35 +572,12 @@ func fileCreate(ctx context.Context, d *schema.ResourceData, m interface{}) diag
 			}...)
 		}
 
-		// the temp directory is used to store the file on the node before moving it to the datastore
-		// will be created if it does not exist
-		tempFileDir := fmt.Sprintf("/tmp/tfpve/%s", uuid.NewString())
-
-		err = capi.SSH().NodeUpload(ctx, nodeName, tempFileDir, request)
+		err = capi.SSH().NodeStreamUpload(ctx, nodeName, *datastore.Path, request)
 		if err != nil {
 			diags = append(diags, diag.FromErr(err)...)
 			return diags
 		}
 
-		// handle the case where the file is uploaded to a subdirectory of the datastore
-		srcDir := tempFileDir
-		dstDir := *datastore.Path
-
-		if request.ContentType != "" {
-			srcDir = tempFileDir + "/" + request.ContentType
-			dstDir = *datastore.Path + "/" + request.ContentType
-		}
-
-		_, err := capi.SSH().ExecuteNodeCommands(ctx, nodeName, []string{
-			// the `mv` command should be scoped to the specific directories in sudoers!
-			fmt.Sprintf(`%s; try_sudo "mv %s/%s %s/%s" && rmdir %s && rmdir %s || echo`,
-				ssh.TrySudo,
-				srcDir, *fileName,
-				dstDir, *fileName,
-				srcDir,
-				tempFileDir,
-			),
-		})
 		if err != nil {
 			if matches, e := regexp.MatchString(`cannot move .* Permission denied`, err.Error()); e == nil && matches {
 				return diag.FromErr(ssh.NewErrUserHasNoPermission(capi.SSH().Username()))

